docs/spec_draft.md

```diff
@@ -341,6 +351,168 @@ only exists to establish data dependencies from `result` to `inputs`.
+operation without getting interfered by execution instances from other groups.
+The number of groups represents the number of `stablehlo.all_to_all` operations
+executed in parallel.
+
+Given,
+  * `num_pids`: Number of partition ids.
+  * `num_rids`: Number of replica ids.
+  * `all_partions_ids`: { pid : pid $\in$ [0, `num_pids`) }.
+  * `all_replica_ids`: { rid : rid $\in$ [0, `num_rids`) }.
+  * `size_subgroup`: size of each subgroup in `replica_groups`.
+  * `num_subgroup`: number of subgroups in `replica_groups`.
+
+The formation of groups is defined as follows:
+   * Number of groups formed: `num_subgroup` * `num_pids`.
+
+   * Size of each group: `size_subgroup`.
+
+   * The formation of `groups` can be demonstrated, using Python-like syntax, as
+     follows:
+     ```python
+     all_replica_ids = [rid for rid in range(num_rids)]
+     all_partition_ids = [pid for pid in range(num_pids)]
+     groups = []
+
+     if len(replica_groups) == 0:
+       replica_groups = [all_replica_ids]
+
+     for replica_group in replica_groups:
+       for pid in all_partition_ids:
+         sub_group = []
+         for rid in replica_group:
+           sub_group.append(ei(rid, pid))
+         groups.append(sub_group)
+     return groups
+     ```
+
+  * For example, assuming `num_pids = 2` and
+      `replica_groups = [[0, 2], [1, 3]]`:
+      `groups` formed: `[ei(0, 0), ei(2, 0)], [ei(0, 1), ei(2, 1)], [ei(1, 0), ei(3, 0)], and [ei(1, 1), ei(3, 1)]`.
+
+For each group `G` $\in$ `groups`, the operation can be described as:
+  * Each `operand`, corresponding to an execution instance in `G`,
+    is split into `split_count` number of blocks, described below using
+    Python-like syntax.
+    ```python
+    blocks = [
+        slice(
+          operand=operand,
+          start_indices=[s0, s1, ..., sR-1],
+            # where
+            #  - sk = 0 if k != split_dimension
+            #  - sk = j * dim(operand) // split_count, if k == split_dimension
+            #  - R = rank(operand)
+          limit_indices=[l0, l1, ..., lR-1],
+            # where
+            #   - lk = dim(operand, k) if k != split_dimension
+            #   - lk = dim(operand, split_dimension) // split_count, if k == split_dimension
+          strides=[t0, t1, ..., tR-1]
+            # where
+            #   - lk = 0 if k != split_dimension
+            #   - lk = dim(operand, split_dimension) // split_count, if k == split_dimension
+        ) for j in range(split_count)]
+    ```
+  * `blocks` are scattered to all the execution instances in `G` such that
+    `blocks[i]` is send to ith execution instance, for i $\in$ G.
+  * Each execution instance in `G` concatenates the received
+    blocks along the `concat_dimension` to produce `result` which is given by
+    `result = concatenate(received_blocks, concat_dimension)`, where
+    `received_blocks` = { received blocks sorted based on the order of execution
+    instances in `G` }.
```

Comment:
Similarly to https://github.com/openxla/stablehlo/pull/503/files#r1032836208, I believe that these PRs can be considerably simplified if we formalize parallel execution semantics beyond the sketch in the "Replica and Partition of a program" section.

Using https://github.com/openxla/stablehlo/pull/634, which provides such a formalization (several pieces of the formalization are TBD, but I think that it's already pretty useful), the "Semantics" section of this op can be compressed to just a few paragraphs:

```
### Semantics

<img align="center" src="spec_draft/all_to_all.svg" />

Within each process group in the StableHLO grid, splits the values of the
`operand` tensor along `split_dimension` into parts, scatters the split parts 
between the processes, concatenates the scattered parts along 
`concat_dimension` and produces a `result` tensor.

The operation splits the StableHLO grid into process groups using the 
`cross_replica(replica_groups)` strategy.

Afterwards, within each `process_group`:
  * ```
    split_parts<sender> = [
        slice(
          operand=operand<sender>,
          start_indices=[s0, s1, ..., sR-1],
            # where
            #  - sj = 0 if j != split_dimension
            #  - sj = i * dim(operand) // split_count, if j == split_dimension
            #  - R = rank(operand)
          limit_indices=[l0, l1, ..., lR-1],
            # where
            #   - lj = dim(operand, j) if j != split_dimension
            #   - lj = dim(operand, split_dimension) // split_count, if j == split_dimension
          strides=[1, ..., 1]
        ) for i in range(split_count)
     ]
     ``` for all `sender` in `process_group`.
  * `scattered_parts<receiver> = [split_parts<sender>[receiver_index] for sender in process_group]` 
    where `receiver_index = index_of(receiver, process_group)`.
  * `result<process> = concatenate(scattered_parts<process>, concat_dimension).
```

Feedback:
  1. See https://github.com/openxla/stablehlo/pull/503/files#r1032836208.
  2. Strides are supposed to be `[1, ..., 1]`.
  3. The `value<process>` syntax allows to precisely formalize the prose in the bullets using formulas.

Reply:
Thanks!
(2) Thanks for the catch. 

I think I still have a problem with  limit_indices: It indicates the end indices, not the increment
fixed it. PTAL.